jobs/audit_jobs.py

```diff
@@ -129,3 +197,91 @@ def expand(self, models_of_kind):
             | 'Apply %s on %s' % (f.__name__, self._kind) >> beam.ParDo(f())
             for f in self._do_fn_types
         )
+
+
+class GetExistingModelKeyCounts(beam.PTransform):
+    """Returns PCollection of (key, count) pairs for each input model."""
+
+    def __init__(self, kind):
+        """Initializes the PTransform.
+
+        Args:
+            kind: str. The kind of model this PTransform will receive.
+        """
+        super(GetExistingModelKeyCounts, self).__init__(
+            label='Generate (key, count)s for all existing %ss' % kind)
+        self._kind = kind
+
+    def expand(self, models_of_kind):
+        """Returns a PCollection of (key, count) pairs for each input model.
+
+        Args:
+            models_of_kind: PCollection. The input models.
+
+        Returns:
+            PCollection. The (ModelKey, int) pairs correponding to the input
+            models and their counts (always 1).
+        """
+        return (
+            models_of_kind
+            | 'Generate (key, count) for %ss' % self._kind >> beam.Map(
+                lambda model: (ModelKey.from_model(model), 1))
+        )
+
+
+class GetMissingModelKeyErrors(beam.PTransform):
+    """Returns PCollection of (key, error) pairs for each referenced model."""
+
+    def __init__(self, kind):
+        """Initializes the PTransform.
+
+        Args:
+            kind: str. The kind of model this PTransform will receive.
+        """
+        super(GetMissingModelKeyErrors, self).__init__(
+            label='Generate (key, error)s from the ID properties in %s' % kind)
+        self._id_property_targets = ID_PROPERTY_TARGETS_BY_MODEL_KIND[kind]
+
+    def expand(self, models_of_kind):
+        """Returns PCollections of (key, error) pairs targeted by the models.
+
+        Args:
+            models_of_kind: PCollection. The input models.
+
+        Returns:
+            iterable(PCollection). The (ModelKey, ModelRelationshipError) pairs
+            corresponding to the models targeted by the ID properties on the
+            input models, and the error that should be reported when they are
+            missing.
+        """
+        return (
+            models_of_kind
+            | 'Generate errors from %s' % id_property >> beam.FlatMap(
+                self._generate_missing_key_errors, id_property, target_kinds)
+
+            for id_property, target_kinds in self._id_property_targets
+        )
+
+    def _generate_missing_key_errors(self, model, id_property, target_kinds):
+        """Yields all model keys referenced by the given model's properties.
+
+        Args:
+            model: Model. The input model.
+            id_property: str. The property that holds the ID(s) of the targeted
```

Comment:
I don't actually think that's anymore helpful, what else would an ID property hold except the ID of a related model?

Reply:
Hm -- to your question, it could contain the ID of the model itself. That's a reasonable interpretation of "id property".

I think it's ok to use the term "IDs", we already do so in the codebase when talking about GAE identifiers.

Reply:
Err I'm confused, then you still don't want to use ID Property? Also, re "it could contain the ID of the model itself", that seems fine to me and matches the definition of any name we choose, right?

Reply:
Oh I meant that it's reasonable to define "id property" as "a field that contains the id of its model instance". This is in response to your question about "what else would an ID property hold" and explaining why the "related model" prefix is useful.

I think "id" should be in the name somehow (because that's literally what the field contains), but "model_reference_id_property" feels too long. So I suppose, despite the concern above, "id_property" might be best after all.

I think I would suggest that:

- when thinking only about the type of the property in isolation, IdProperty suffices (analogous to StringProperty etc.)
- when using the property in the context of references to other models, have something in the code snippet or docstring that explains that, or use a local variable name that conveys that.

I agree it's a bit hard to come up with good naming here.

Reply:
OK, how are these names?